src/cargo/ops/cargo_compile.rs

```diff
@@ -560,9 +564,16 @@ fn generate_targets<'a>(
             } => {
                 if lib {
                     if let Some(target) = pkg.targets().iter().find(|t| t.is_lib()) {
+                        if build_config.mode == CompileMode::Doctest && !target.doctestable() {
+                            bail!(
+                                "doc tests are not supported for crate type(s) `{}` in package `{}`",
+                                target.rustc_crate_types().join(", "),
+                                pkg.name()
+                            );
+                        }
```

Comment:
Will it be possible to issue this error earlier, during manifest parsing phase?

Reply:
I don't think so, most of that code doesn't know the build mode.  The only place I can think of is `commands/test.rs`, but it would need to duplicate the iterations over packages and targets.

Reply:
Hm, you are correct. Still, something things a bit wrong about this error. If you have `[lib] doctest = false` and run `carogo test --doc`, you won't get any errors, the test will be silently skipped. If you have incompatible crate type, you'll get a loud error.

I think that we might want to unify the behavior for both cases perhaps?

Reply:
Heh, I initially wrote a warning for that, but then decided to drop it.  I was concerned that it was inconsistent.  It won't be skipped, it will actually run the doctest.   Today, if you use a target filter, it always ignores the enabled flags (`test`, `bench`, `doctest`, and `doc`).  I wasn't sure if that was intentional or not, but I didn't want to make it inconsistent or change the existing behavior.  This is related to https://github.com/rust-lang/cargo/pull/5725#issuecomment-405277792.  If you do want to change that behavior, I think it should be changed for all the target types (I feel like inconsistent behavior can be confusing).